src/libraries/System.Private.CoreLib/src/System/Decimal.DecCalc.cs

```diff
@@ -479,24 +546,59 @@ private static uint Div128By96(ref Buf16 bufNum, ref Buf12 bufDen)
             /// <returns>Returns highest 32 bits of product</returns>
             private static uint IncreaseScale(ref Buf12 bufNum, uint power)
             {
-                ulong tmp = Math.BigMul(bufNum.U0, power);
+#if TARGET_64BIT
+                ulong hi64 = Math.BigMul(bufNum.Low64, power, out ulong low64);
+                bufNum.Low64 = low64;
+                hi64 = Math.BigMul(bufNum.U2, power) + hi64;
+                bufNum.U2 = (uint)hi64;
+                return (uint)(hi64 >> 32);
+#else
+                ulong tmp = (ulong)bufNum.U0 * power;
                 bufNum.U0 = (uint)tmp;
                 tmp >>= 32;
-                tmp += Math.BigMul(bufNum.U1, power);
+                tmp += (ulong)bufNum.U1 * power;
                 bufNum.U1 = (uint)tmp;
                 tmp >>= 32;
-                tmp += Math.BigMul(bufNum.U2, power);
+                tmp += (ulong)bufNum.U2 * power;
                 bufNum.U2 = (uint)tmp;
                 return (uint)(tmp >> 32);
+#endif
             }
 
-            private static void IncreaseScale64(ref Buf12 bufNum, uint power)
+            /// <summary>
+            /// Multiply the two numbers. The result overwrite the input.
+            /// </summary>
+            /// <param name="bufNum">buffer</param>
+            /// <param name="power">Scale factor to multiply by</param>
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private static void IncreaseScale(ref Buf16 bufNum, uint power)
             {
-                ulong tmp = Math.BigMul(bufNum.U0, power);
+#if TARGET_64BIT
+                ulong hi64 = Math.BigMul(bufNum.Low64, power, out ulong low64);
+                bufNum.Low64 = low64;
+                bufNum.High64 = Math.BigMul(bufNum.U2, power) + (nuint)hi64;
+#else
+                ulong tmp = (ulong)bufNum.U0 * power;
                 bufNum.U0 = (uint)tmp;
                 tmp >>= 32;
-                tmp += Math.BigMul(bufNum.U1, power);
+                tmp += (ulong)bufNum.U1 * power;
+                bufNum.U1 = (uint)tmp;
+                tmp >>= 32;
+                tmp += (ulong)bufNum.U2 * power;
                 bufNum.High64 = tmp;
```

Comment:
This can be replaced with the much more concise unsafe code if you think it would be a god change.
The buffer should be on the stack so it should be safe? 

It would reduce the number of lines with 7 at the cost of 6% slower division when divisor is 96 bit (x86).

It seems to me that the performace drop (probably mostly due to extra stack spills) is something the jit should be able to eliminate in the future

```suggestion
                bufNum.U3 = IncreaseScale(ref *(Buf12*)Unsafe.AsPointer(ref bufNum), power);
```

Reply:
also needs to make method unsafe to do this

Reply:
Note: this overload did make a significant improvement to 64bit speed (probably due to store to load forwarding not working without)

Reply:
> `bufNum.U3 = IncreaseScale(ref *(Buf12*)Unsafe.AsPointer(ref bufNum), power);`

This is a GC hole. You have to use `Unsafe.As` for this.

Reply:
Looks like in this case it's sort of safe, because `Div96By32` is called from one place where the buffer is on the stack, but it's better to avoid such patterns with a complex contract (the struct is not even byref).

Reply:
See https://github.com/dotnet/runtime/issues/99144, we're trying to remove uses of `Unsafe.AsPointer`, not add more where `As` can be used 😆

Reply:
So `bufNum.U3 = IncreaseScale(ref Unsafe.As<Buf16, Buf12>(ref bufNum), power);` should be safe to use ?

I've updated the suggestion to use that instead, it also seems to generate better code (ret is as offset 91 instead of a7) with less stack use which is good. I have not made any new measurements, but the assembly for `IncreaseScale (Buf16&, UInt32)` is now almost identical to the code without the call (at least before any inlining) so its should not have the same performance impact as the pointer version.

I will apply the modified suggestion with `As` and can always revert the change if desired